src/corporacreator/corpus.py

```diff
@@ -78,18 +78,30 @@ def _post_process_valid_data(self):
         )
         self.valid = self.valid.sort_values(["user_sentence_count", "client_id"])
         valid = self.valid.groupby("sentence").head(self.args.duplicate_sentence_count)
+        
         valid = valid.sort_values(["user_sentence_count", "client_id"], ascending=False)
         valid = valid.drop(columns="user_sentence_count")
         self.valid = self.valid.drop(columns="user_sentence_count")
         # Determine train, dev, and test sizes
         train_size, dev_size, test_size = self._calculate_data_set_sizes(len(valid))
         # Split into train, dev, and test datasets
-        self.train = valid.iloc[0:train_size]
-        self.dev = valid.iloc[train_size : train_size + dev_size]
-        self.test = valid.iloc[
-            train_size + dev_size : train_size + dev_size + test_size
-        ]
-        # TODO: Make sure users are in train, dev, xor test
+        continous_client_index, uniques = pd.factorize(valid["client_id"])
+        valid["continous_client_index"] = continous_client_index
+        train = pd.DataFrame(columns=valid.columns)
+        dev = pd.DataFrame(columns=valid.columns)
+        test = pd.DataFrame(columns=valid.columns)
+
+        for i in range(max(continous_client_index), -1, -1):
+            if len(test) + len(valid[valid["continous_client_index"] == i]) <= test_size:
+                test = pd.concat([test, valid[valid["continous_client_index"] == i]])
+            elif len(dev) + len(valid[valid["continous_client_index"] == i]) <= dev_size:
+                dev = pd.concat([dev, valid[valid["continous_client_index"] == i]])
+            else:
```

Comment:
Shouldn't there be a test here to make sure that not too many elements are added to the train?

Something like the following:

```python
            elif len(train) + len(valid[valid["continous_client_index"] == i]) <= train_size:
```

Reply:
I honestly thought for train dataset: "the more, the merrier" :rofl: 

The in this PR proposed processing adds sentences / rows in batches-per-client_id to each of the datasets. So if we add 
`elif len(train) + len(valid[valid["continous_client_index"] == i]) <= train_size:`
to it, it may occur, that the last batch of data to add exceeds `train_size`, which in turn can result in having less than the intended train_size in the training dataset. Or in other words, if the last batch size is 2 but there is only train_size -1 space in the training dataset left, the we would end up with train_size-1 rows.

So I suggest, we change the line:
`self.test = test.drop(columns="continous_client_index")`
to
`self.test = test[:train_size].drop(columns="continous_client_index")`

That way, we maybe add more rows to the dataset than we need but cut the length of the dataset accorting to train_size.

Reply:
I like the

```python
self.test = test[:train_size].drop(columns="continous_client_index")
```

idea!